File: backend/app/schemas/business_unit.py
"""Pydantic schemas for Business Unit endpoints."""

from datetime import datetime
from uuid import UUID
from typing import Optional, List

from pydantic import BaseModel, Field


# ===== Base Schemas =====

class BusinessUnitBase(BaseModel):
    """Base business unit schema."""
    name: str = Field(..., min_length=1, max_length=255)
    code: str = Field(..., min_length=1, max_length=50)
    description: Optional[str] = None


# ===== Request Schemas =====

class BusinessUnitCreate(BusinessUnitBase):
    """Schema for creating a business unit."""
    parent_id: Optional[UUID] = None
    head_user_id: Optional[UUID] = None


class BusinessUnitUpdate(BaseModel):
    """Schema for updating a business unit."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    description: Optional[str] = None
    parent_id: Optional[UUID] = None
    head_user_id: Optional[UUID] = None
    is_active: Optional[bool] = None


# ===== Response Schemas =====

class BusinessUnitSummary(BaseModel):
    """Minimal business unit info for embedding in other responses."""
    id: UUID
    name: str
    code: str
    is_active: bool

    class Config:
        from_attributes = True


class BusinessUnitResponse(BusinessUnitBase):
    """Schema for business unit response."""
    id: UUID
    tenant_id: UUID
    parent_id: Optional[UUID] = None
    head_user_id: Optional[UUID] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class BusinessUnitWithHierarchy(BusinessUnitResponse):
    """Business unit with parent and children info."""
    parent: Optional[BusinessUnitSummary] = None
    children: List[BusinessUnitSummary] = []
    full_path: str

    class Config:
        from_attributes = True


class BusinessUnitListResponse(BaseModel):
    """Schema for paginated business unit list."""
    items: List[BusinessUnitResponse]
    total: int
    page: int
    page_size: int
    pages: int


class BusinessUnitTree(BaseModel):
    """Business unit with nested children for tree view."""
    id: UUID
    name: str
    code: str
    description: Optional[str] = None
    is_active: bool
    head_user_id: Optional[UUID] = None
    children: List["BusinessUnitTree"] = []

    class Config:
        from_attributes = True


# Allow forward reference for recursive type
BusinessUnitTree.model_rebuild()

File: frontend/src/types/business-unit.ts
// Business Unit types

export interface BusinessUnit {
  id: string
  tenant_id: string
  name: string
  code: string
  description?: string
  parent_id?: string
  head_user_id?: string
  industry_profile_id?: string
  effective_profile_name?: string
  is_active: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  config_overrides?: Record<string, any> | null
  created_at: string
  updated_at: string
}

export interface BusinessUnitTree {
  id: string
  name: string
  code: string
  description?: string
  is_active: boolean
  head_user_id?: string
  industry_profile_id?: string
  effective_profile_name?: string
  children: BusinessUnitTree[]
}

export interface BusinessUnitCreate {
  name: string
  code: string
  description?: string | null
  parent_id?: string | null
  head_user_id?: string | null
  industry_profile_id?: string | null
}

export interface BusinessUnitUpdate {
  name?: string
  code?: string
  // null = explicitly clear (undefined keys are dropped by axios and the
  // backend's exclude_unset treats them as "no change")
  description?: string | null
  parent_id?: string | null
  head_user_id?: string | null
  industry_profile_id?: string | null
  is_active?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  config_overrides?: Record<string, any>
}

export interface BusinessUnitListResponse {
  items: BusinessUnit[]
  total: number
  page: number
  page_size: number
  pages: number
}
